extension/popup/popup.jsx

```diff
@@ -0,0 +1,744 @@
+            // Probably superfluous, since this is called in onProcessing:
+            browser.runtime.sendMessage({ type: "microphoneStopped" })
+            clearInterval(recorderIntervalId)
+            this.setCurrentState("success")
+            if (json === null) {
+                // It was cancelled
+                return
+            }
+            browser.runtime.sendMessage({
+                type: "addTelemetry",
+                properties: { transcriptionConfidence: json.data[0].confidence },
+            })
+            this.setTranscript(json.data[0].text)
+            executedIntent = true
+            browser.runtime.sendMessage({
+                type: "runIntent",
+                text: json.data[0].text,
+            })
+        }
+        recorder.onError = error => {
+            browser.runtime.sendMessage({ type: "microphoneStopped" })
+            log.error("Got recorder error:", String(error), error)
+            this.setCurrentState("error")
+            // TODO: No error message is shown yet. Show one?
+            //this.setState({ error })
+            clearInterval(recorderIntervalId)
+        }
+        recorder.onProcessing = () => {
+            browser.runtime.sendMessage({ type: "microphoneStopped" })
+            this.setCurrentState("processing")
+        }
+        recorder.onNoVoice = () => {
+            browser.runtime.sendMessage({ type: "microphoneStopped" })
+            log.debug("Closing popup because of no voice input")
+            window.close()
+        }
+        recorder.startRecording()
+    }
+
+    handleMessage = (message) => {
+        switch(message.type) {
+            case "closePopup": {
+                closePopup(message.time)
+                break
+            }
+            case "displayFailure": {
+                this.setCurrentState("error")
+                if (message.message) {
+                    this.setState({ error: message.message })
+                }
+                break
+            }
+            case "displayText": {
+                this.setState({ displayText: message.message })
+                overrideTimeout = TEXT_TIMEOUT
+                break
+            }
+            case "displayAutoplayFailure": {
+                this.setErrorMessage("Please enable autoplay on this site for a better experience")
+                this.setState({ displayAutoplay: true })
+                break
+            }
+            case "showSearchResults": {
+                this.setState({ search: message })
+                this.setCurrentState("searchResults")
+                return Promise.resolve(true)
+            }
+            default:
+                break
+        }
+        return
+    }
+
+    updateExamples = async () => {
+        const suggestions = await browser.runtime.sendMessage({
+            type: "getExamples",
+            number: 3,
+        })
+
+        this.setState({ suggestions })
+    }
+
+    setCurrentState = (currentState) => {
```

Comment:
I'm not sure this is the right pattern to be using with React, or at least not in its entirety. I think it makes sense for a couple cases:

1. Waiting for voice
2. Waiting for typing (once typing has begun)
3. After a command is heard/entered.

The third case has a bunch of cases, like a card, an error message, some text, etc. But those can each be properties, and each displayed if the data is present.

That said, it's probably best to leave this as it is for now without trying to change everything on the first pass. This is already a big patch!